### PR TITLE
[FLINK-2793] [dashboard] Redirect to leader in non-standalone mode

### PR DESCRIPTION
**Pull request by uce**
- Redirects HTTP requests to the web fronted of the leading job manager if not associated with leading job manager
- Splits up the start up of the job manager and web frontend (see `JobManager#runJobManager`) to first start the web fronted (binding to the port), then the job manager actor, then the web fronted with the address of the associated job manager.
- The port of the web frontend to redirect is determined via a newly introduced `JobManagerMessage`
- The associated job manager is used to determine whether the leader notification addresses the associated job manager or another one.
- Adds web front end tests via HttpTestClient (moved the generated files to the src folder for this) 

This breaks the leader notification abstraction in the sense that the job manager location is not transparent anymore. The reason for this is that the execution graph structures are not serializable (a proper fix would address this instead of redirecting). I have not integrated this with the TestingCluster.

### CODE REVIEW

**Comment by rmetzger** [comment]
+1 The change looks good to merge. (maybe we should add the JIRA id when merging)

**Comment by uce** [comment]
I would squash all commits and add the JIRA. Have you tried it out?

I would like to hear @tillrohrmann's take on this "abstraction" breaking behaviour before merging it.

**Comment by rmetzger** [comment]
No, I didn't try it.

**Comment by tillrohrmann** [comment]
I already opened the PR. Will review it asap.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 1]
```diff
@@ -129,6 +129,13 @@ class JobManager(
   var leaderSessionID: Option[UUID] = None
 
   /**
+   * The port of the web monitor as configured. Make sure that it is actually configured before
+   * starting the JobManager.
+   */
+  val webMonitorPort : Integer = flinkConfiguration.getInteger(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: Why not using Scala's `Int` type here?

Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: Maybe we can add a comment which says that this is gonna be kicked out once we have made the `ExecutionGraph` properly serializable or so...

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 2]
```diff
@@ -1761,11 +1803,8 @@ object JobManager {
    * @param config The configuration for the runtime monitor.
    * @param leaderRetrievalService Leader retrieval service to get the leading JobManager
    */
-  def startWebRuntimeMonitor(
-      config: Configuration,
-      leaderRetrievalService: LeaderRetrievalService,
-      actorSystem: ActorSystem)
-    : WebMonitor = {
+  def startWebRuntimeMonitor(config: Configuration, leaderRetrievalService: LeaderRetrievalService,
+    actorSystem: ActorSystem) : WebMonitor = {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: Line breaks not very scalaesque

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 3]
```diff
@@ -1773,8 +1812,7 @@ object JobManager {
         .asSubclass(classOf[WebMonitor])
 
       val ctor: Constructor[_ <: WebMonitor] = clazz.getConstructor(classOf[Configuration],
-        classOf[LeaderRetrievalService],
-        classOf[ActorSystem])
+        classOf[LeaderRetrievalService], classOf[ActorSystem])
```
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: Line breaks not very scalaesque

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java [hunk 4]
```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
...
+	}
 
 	@Override
 	public void notifyLeaderAddress(String leaderAddress, UUID leaderSessionID) {
 		if (leaderAddress != null && !leaderAddress.equals("")) {
 			try {
-				ActorRef jobManager = AkkaUtils.getActorRef(
-						leaderAddress,
-						actorSystem,
+				ActorRef jobManager = AkkaUtils.getActorRef(leaderAddress, actorSystem,
```
Comment (tillrohrmann) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java]: This is a blocking call which will block the `NodeCache's` thread which calls the `NodeCacheListener`. This should be avoided by using `AkkaUtils.getActorRefFuture` and a `Promise`.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java [hunk 5]
```diff
@@ -70,28 +104,68 @@ public ActorGateway getArchiveGateway() {
...
+					// We need to redirect to the leader -.-
+					//
+					// This is necessary currently, because many execution graph structures are not
+					// serializable. The proper solution here is to have these serializable and
+					// transparently work with the leading job manager instead of redirecting.
+					Future<Object> portFuture = jobManagerGateway
+							.ask(JobManagerMessages.getRequestWebMonitorPort(), timeout);
+
+					JobManagerMessages.ResponseWebMonitorPort portResponse =
+							(JobManagerMessages.ResponseWebMonitorPort) Await.result(portFuture, timeout);
```
Comment (tillrohrmann) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java]: Again a blocking call which is bad. Better to use futures to circumvent this problem.

**Comment by tillrohrmann** [comment]
The PR looks good. I think we cannot do much about it being a temporary workaround for the limitation that our `ExecutionGraph` is not serializable. Once we have the resources this should be addressed, though.

I've got some minor comments. The main point is the blocking call in the `notifyLeaderAddress` of the `JobManagerArchiverRetriever`. This should be fixed imo.

Moreover, I think I've found a race condition. The `RuntimeMonitorHandler` checks whether the current leader is its own or a remote one. If it's the local JM, then the respective RequestHandler is called. Within the `RequestHandler`, such as `CurrentJobsOverviewHandler` the `JobManagerArchiverRetriever` is called again. If now the leader has changed to be a remote JM, then we'll have a problem. This could be fixed as well.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java (hunk 4 above)
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java]: Good catch

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala (hunk 3 above)
Comment (uce) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: I was going for Kafkaesque... :D Just kidding, will change it.

**Comment by tillrohrmann** [comment]
What you can actually do to resolve the issue with the `RequestHandler` is to statically assign them the local JM's address instead of providing a `JobManagerArchiveRetriever`. Since they are only called when the local JM is the leader, they don't have to look up the current leader address.

**Comment by uce** [comment]
Yes, good point. I will do this now.

**Comment by uce** [comment]
OK. I've addressed your comments in the last two commits. Thanks again!

- The associated local leader gateway is only resolved once and getGateway will only return this one (new leader notifications don't change this one and only affect the redirect address)
- The handlers don't block on the redirect address, but just read the current redirect address.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java [hunk 6]
```diff
@@ -47,51 +60,160 @@
...
+		this.timeout = checkNotNull(timeout);
+	}
+
+	/**
+	 * Associates this instance with the job manager identified by the given URL.
+	 *
+	 * <p>This has to match the URL retrieved by the leader retrieval service. In tests setups you
+	 * have to make sure to use the correct type of URLs.
+	 */
+	public void setJobManagerAkkaUrlAndRetrieveGateway(String jobManagerAkkaUrl) throws Exception {
```
Comment (tillrohrmann) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/JobManagerArchiveRetriever.java]: Actually, I meant to get rid of this extra redirect logic here.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/RuntimeMonitorHandler.java [hunk 7]
```diff
@@ -44,27 +47,40 @@
...
+	public RuntimeMonitorHandler(RequestHandler handler, JobManagerArchiveRetriever retriever) {
+		this.retriever = checkNotNull(retriever);
+		this.handler = checkNotNull(handler);
 		this.contentType = (handler instanceof RequestHandler.JsonResponse) ? "application/json" : "text/plain";
 	}
 	
 	@Override
 	protected void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Exception {
+		// Redirect to leader if necessary
+		String redirectAddress = retriever.getRedirectAddress();
```
Comment (tillrohrmann) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/RuntimeMonitorHandler.java]: Get rid of this `getRedirectAddress` method here and simply ask for the current leader. Then compare the leader address with the local job manager address to decide whether you have to redirect or not.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java [hunk 8]
```diff
@@ -150,97 +148,97 @@ else if (flinkRoot != null) {
...
 
 			// job manager configuration
-			.GET("/jobmanager/config", handler(new JobManagerConfigHandler(config)))
+			.GET("/jobmanager/config", handler(new JobManagerConfigHandler(config), retriever))
 
 			// overview over jobs
-			.GET("/joboverview", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, true)))
-			.GET("/joboverview/running", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, false)))
-			.GET("/joboverview/completed", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, false, true)))
+			.GET("/joboverview", handler(new CurrentJobsOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT, true, true), retriever))
```
Comment (tillrohrmann) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: Start the `RequestHandler` with the JobManager address of the local JobManager. They will only be called from the `RuntimeMonitorHandler` if the local JM is the leader.

**Comment by tillrohrmann** [comment]
I've added some more comments. Ping me if it's too superficial.

**Comment by uce** [comment]
It's not superficial at all. It's cleaner, because the retriever is independent of the redirect logic then.

I've decided against it though in order to keep the handlers cleaner (there only two different kinds atm though, so it would be OK to add it) and because the associated job manager URL is only known after start() is called, so I would have to update all handlers lazily after that.

This is due to the start up ordering of
1. start web server (bind to port)
2. get web server port and start job manager with web server port configured
3. set job manager address at web server

This ordering allows both components to pick random ports.

I just didn't want to change too much code this at this point, because I genuinely hope that we will make the execution graph serialisable after the release and get back to the old solution. But I'm also happy to change it if you like. Just ping me.

**Comment by rmetzger** [comment]
Are there any blockers to this PR. Otherwise, I'd like to have it merged rather soon.

**Comment by tillrohrmann** [comment]
Have fixed a possible race condition in `JobManagerRetriever` and rebased on the latest master. If Travis gives green light, I'll merge it.

**Comment by rmetzger** [comment]
Great ;)

**Comment by tillrohrmann** [comment]
Oh I forgot. I still want to test it on Yarn.

**Pull request closed without merging**